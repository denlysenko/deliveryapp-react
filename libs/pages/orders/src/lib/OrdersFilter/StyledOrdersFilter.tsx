import styled from 'styled-components';

export const StyledOrdersFilter = styled.div`
  .order-filter {
    .p-md-6 {
      padding: 0;
    }

    .p-inputgroup {
      @media (max-width: 767px) {
        margin-bottom: 16px;
      }

      input {
        @media (min-width: 641px) {
          margin-right: 16px;
        }
      }
    }
  }
`;
