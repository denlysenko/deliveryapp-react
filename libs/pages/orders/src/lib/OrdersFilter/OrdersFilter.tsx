import React, { useState, useCallback } from 'react';

import { SelectItem } from 'primereact/api';
import { Dropdown } from 'primereact/dropdown';
import { InputText } from 'primereact/inputtext';

import { debounce, isNil } from 'lodash-es';

import { OrdersFilter as IOrdersFilter } from '@deliveryapp/data-access';

import { StyledOrdersFilter } from './StyledOrdersFilter';

export const options: SelectItem[] = [
  {
    label: 'Order number',
    value: 'id'
  },
  {
    label: 'Cargo Name',
    value: 'cargoName'
  },
  {
    label: 'From',
    value: 'cityFrom'
  },
  {
    label: 'To',
    value: 'cityTo'
  }
];

export interface OrdersFilterProps extends Pick<IOrdersFilter, 'filter'> {
  handleFilterChange: (orderFilter: IOrdersFilter['filter']) => void;
}

export const OrdersFilter: React.FC<OrdersFilterProps> = ({
  filter,
  handleFilterChange
}) => {
  const [searchTerm, setSearchTerm] = useState(
    !isNil(filter) ? Object.values(filter)[0] ?? '' : ''
  );

  const [criteria, setCriteria] = useState(
    !isNil(filter)
      ? Object.keys(filter)[0] ?? options[0].value
      : options[0].value
  );

  // https://github.com/facebook/react/issues/1360#issuecomment-533847123
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const debounceHandleFilterChange = useCallback(
    debounce(
      (orderFilter: IOrdersFilter['filter']) => handleFilterChange(orderFilter),
      500
    ),
    []
  );

  const handleInputChange = (event: React.FormEvent<HTMLInputElement>) => {
    const value = event.currentTarget.value;
    setSearchTerm(value);
    debounceHandleFilterChange({ [criteria]: value });
  };

  const handleDropdownChange = (event: { value: string }) => {
    const value = event.value;

    setCriteria(value);

    if (searchTerm) {
      handleFilterChange({ [value]: searchTerm });
    }
  };

  return (
    <StyledOrdersFilter>
      <div className="order-filter p-grid">
        <div className="p-inputgroup p-col-12 p-md-6">
          <span className="p-inputgroup-addon">
            <i className="fa fa-search"></i>
          </span>
          <InputText
            data-testid="searchTerm"
            type="text"
            placeholder="Search..."
            value={searchTerm}
            onChange={handleInputChange}
          />
        </div>
        <div className="p-col-12 p-md-6">
          <Dropdown
            options={options}
            value={criteria}
            onChange={handleDropdownChange}
          />
        </div>
      </div>
    </StyledOrdersFilter>
  );
};
