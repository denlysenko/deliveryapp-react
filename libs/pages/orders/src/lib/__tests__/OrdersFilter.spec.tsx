/* eslint-disable @typescript-eslint/no-non-null-assertion */
import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import { options, OrdersFilter } from '../OrdersFilter/OrdersFilter';

jest.mock('lodash', () => ({
  ...require.requireActual('lodash'),
  debounce: jest.fn((fn) => fn)
}));

const handleFilterChange = jest.fn();

describe('OrdersFilter', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render successfully', () => {
    const { baseElement } = render(
      <OrdersFilter handleFilterChange={handleFilterChange} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should render options', () => {
    render(<OrdersFilter handleFilterChange={handleFilterChange} />);
    expect(screen.getAllByRole('option')).toHaveLength(options.length);
  });

  it('should render first option for criteria if filter prop is not present', () => {
    const { container } = render(
      <OrdersFilter handleFilterChange={handleFilterChange} />
    );
    expect(container.querySelector('.p-dropdown-label')).toHaveTextContent(
      options[0].label!
    );
  });

  it('should render first option for criteria if filter prop is empty', () => {
    const { container } = render(
      <OrdersFilter filter={{}} handleFilterChange={handleFilterChange} />
    );
    expect(container.querySelector('.p-dropdown-label')).toHaveTextContent(
      options[0].label!
    );
  });

  it('should render empty string for searchTerm if filter prop is not present', () => {
    render(<OrdersFilter handleFilterChange={handleFilterChange} />);
    expect(screen.getByPlaceholderText(/search/i)).toHaveValue('');
  });

  it('should render empty string for searchTerm if filter prop is empty', () => {
    render(
      <OrdersFilter filter={{}} handleFilterChange={handleFilterChange} />
    );
    expect(screen.getByPlaceholderText(/search/i)).toHaveValue('');
  });

  it('should render searchTerm and criteria from filter', () => {
    const { container } = render(
      <OrdersFilter
        filter={{ [options[2].value]: 'test' }}
        handleFilterChange={handleFilterChange}
      />
    );
    expect(container.querySelector('.p-dropdown-label')).toHaveTextContent(
      options[2].label!
    );
    expect(screen.getByPlaceholderText(/search/i)).toHaveValue('test');
  });

  describe('handleFilterChange', () => {
    it('should not call if criteria has changed, but searchTerm is empty', () => {
      render(
        <OrdersFilter filter={{}} handleFilterChange={handleFilterChange} />
      );

      fireEvent.click(screen.getAllByRole('option')[0]);
      expect(handleFilterChange).toBeCalledTimes(0);
    });

    it('should call if searchTerm has changed', () => {
      render(
        <OrdersFilter filter={{}} handleFilterChange={handleFilterChange} />
      );

      fireEvent.change(screen.getByPlaceholderText(/search/i), {
        target: {
          value: 'test'
        }
      });

      expect(handleFilterChange).toBeCalledTimes(1);
      expect(handleFilterChange).toBeCalledWith({ [options[0].value]: 'test' });
    });

    it('should call if criteria has changed', () => {
      render(
        <OrdersFilter
          filter={{ [options[2].value]: 'test' }}
          handleFilterChange={handleFilterChange}
        />
      );

      fireEvent.click(screen.getAllByRole('option')[1]);

      expect(handleFilterChange).toBeCalledTimes(1);
      expect(handleFilterChange).toBeCalledWith({ [options[1].value]: 'test' });
    });
  });
});
