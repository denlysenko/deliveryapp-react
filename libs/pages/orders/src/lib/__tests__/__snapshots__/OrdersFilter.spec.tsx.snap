// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrdersFilter should render successfully 1`] = `
<body>
  <div>
    <div
      class="sc-AxjAm isfZYR"
    >
      <div
        class="order-filter p-grid"
      >
        <div
          class="p-inputgroup p-col-12 p-md-6"
        >
          <span
            class="p-inputgroup-addon"
          >
            <i
              class="fa fa-search"
            />
          </span>
          <input
            class="p-inputtext p-component"
            placeholder="Search..."
            type="text"
            value=""
          />
        </div>
        <div
          class="p-col-12 p-md-6"
        >
          <div
            class="p-dropdown p-component"
          >
            <div
              class="p-hidden-accessible"
            >
              <input
                aria-haspopup="listbox"
                readonly=""
                type="text"
              />
            </div>
            <div
              class="p-hidden-accessible p-dropdown-hidden-select"
            >
              <select
                aria-hidden="true"
                tabindex="-1"
              >
                <option
                  value=""
                />
                <option
                  value="id"
                >
                  Order number
                </option>
              </select>
            </div>
            <label
              class="p-dropdown-label p-inputtext"
            >
              Order number
            </label>
            <div
              aria-haspopup="listbox"
              class="p-dropdown-trigger"
              role="button"
            >
              <span
                class="p-dropdown-trigger-icon pi pi-chevron-down p-clickable"
              />
            </div>
            <div
              class="p-dropdown-panel p-hidden p-input-overlay"
            >
              <div
                class="p-dropdown-items-wrapper"
                style="max-height: 200px;"
              >
                <ul
                  class="p-dropdown-items p-dropdown-list p-component"
                  role="listbox"
                >
                  <li
                    aria-label="Order number"
                    aria-selected="true"
                    class="p-dropdown-item p-highlight"
                    role="option"
                  >
                    Order number
                  </li>
                  <li
                    aria-label="Cargo Name"
                    aria-selected="false"
                    class="p-dropdown-item"
                    role="option"
                  >
                    Cargo Name
                  </li>
                  <li
                    aria-label="From"
                    aria-selected="false"
                    class="p-dropdown-item"
                    role="option"
                  >
                    From
                  </li>
                  <li
                    aria-label="To"
                    aria-selected="false"
                    class="p-dropdown-item"
                    role="option"
                  >
                    To
                  </li>
                </ul>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
